mlflow/tracing/utils/__init__.py

```diff
@@ -73,13 +73,40 @@ def default(self, obj):
         # representation, so we use dict representation instead.
         # E.g. https://github.com/run-llama/llama_index/blob/29ece9b058f6b9a1cf29bc723ed4aa3a39879ad5/llama-index-core/llama_index/core/chat_engine/types.py#L63-L64
         if is_dataclass(obj):
-            return asdict(obj)
+            try:
+                return asdict(obj)
+            except TypeError:
+                pass
+
+        # Some object has dangerous side effect in __str__ method, so we use class name instead.
+        if not self._is_safe_to_encode_str(obj):
+            return type(obj)
 
         try:
             return super().default(obj)
         except TypeError:
             return str(obj)
 
+    def _is_safe_to_encode_str(self, obj) -> bool:
```

Comment:
note: Sometimes span input also contains a generator, e.g. an formatter takes `StreamingResponse` from the previous LLM step. This check is to avoid consuming the generator due to the weird `__str__` implementation of those data classes..

Reply:
I wonder if those implementations are intentionally consuming the generator or if that's a design oversight. Might be worth giving them a heads up on the behavior :)

Reply:
https://github.com/run-llama/llama_index/blob/f092d90bd5934097f5c166014f5d99a3e07ea999/llama-index-core/llama_index/core/base/response/schema.py#L120-L127

It does look like it's intended.

Reply:
Yea, it's tricky but surely a design choice😐